retention messaging (#422)
adaptyteam/adapty-docs: src/content/docs/version-3.0/retention-messaging-setup.mdx

---

> title: "Set up Retention Messaging"

Reviewer: не хватило ещё объяснения про тоггл переключения между сэндбоксом и продом

как работает это:
After sandbox passes, the user reaches production by turning off the "Sandbox mode" toggle above the table — the production setup card then appears in place of the table. Alternatively, the create-message screen offers a "Configure Production URL" link. The article says "click Configure URL under the production setup" without explaining where that is.

description: "Connect the Retention Messaging realtime URL to Adapty: configure it in sandbox, pass Apple's performance test, then switch to production."
metadataTitle: "Set up Retention Messaging | Adapty Docs"
---

:::link
Main article: [Retention Messaging](retention-messaging.md)
:::

> Before Adapty can show retention messages, Apple needs a way to reach it when a subscriber cancels. You set up that connection here — in sandbox first to test it, then in production.

Reviewer: ```suggestion
Before subscribers can see your retention messages, Apple needs a way to reach Adapty when someone cancels. This article shows how to set up that connection — in sandbox first, then in production.
```

## Before you begin

- Add an [App Store Connect API key](app-store-connection-configuration) to Adapty. Adapty uses it to register the realtime URL and communicate with Apple.

> - Request access to [Apple's Retention Messaging program](https://developer.apple.com/documentation/retentionmessaging) and wait for Apple's approval. Not every app is eligible, so ask early — without approval, Apple rejects the connection.

Reviewer: ```suggestion
- Request access to [Apple's Retention Messaging program](https://developer.apple.com/documentation/retentionmessaging) and wait for Apple's approval. Without approval, Apple rejects the connection.
```

## Connect Adapty in sandbox

You set this up in Apple's sandbox environment first, so you can verify the connection before any real subscriber sees a message.

> In the Adapty Dashboard, go to **Retention Messaging** and click **Configure Sandbox URL**. A setup card opens and guides you through the connection.

Reviewer: ```suggestion
In the Adapty Dashboard, go to **Retention Messaging** and click **Configure Sandbox URL**. A setup card opens with the connection steps
```

{/* screenshot here */}

> On the card, click **Configure URL**. Adapty registers its endpoint with Apple through the App Store Connect API. If App Store Connect already has a different URL set, the card shows **Replace URL** instead — click it to point Apple at Adapty.

Reviewer: ```suggestion
On the card, click **Configure URL**. Adapty registers its endpoint with Apple through the App Store Connect API. If App Store Connect already has a different URL set, the card shows **Replace URL** instead — click it to replace the existing URL with Adapty's.
```

This URL is dedicated to Retention Messaging. It's separate from the App Store Server Notifications you set when connecting your App Store Connect account.

> When it succeeds, Adapty confirms with "Sandbox URL has been configured."

Reviewer: ```suggestion
When the connection succeeds, Adapty confirms with "Sandbox URL has been configured."
```

{/* screenshot here */}

## Test the connection

> Before you can go to production, Apple requires a sandbox test (it calls this a **performance test**). It confirms that Adapty answers Apple's request fast enough — within about 700 ms.

Reviewer: ```suggestion
Before you can go to production, Apple requires a sandbox test (Apple calls this a **performance test**). The test confirms that Adapty answers Apple's request fast enough — within about 700 ms.
```

> The test runs against a sandbox transaction, so first make a purchase with a Sandbox Apple Account, the same way you test purchases.

Reviewer: не хватает ссылочки на статью про тестирование

> On the same card, click **Start Test**. The test takes a few minutes. If it fails, check the connection, then click **Retry Test**.

Reviewer: а что именно проверить? если не знаем, лучше не писать

{/* screenshot here */}

## Switch to production

> You can only connect production after the sandbox test passes — until then, the production step stays disabled. Once it passes, click **Configure URL** under the production setup. Adapty confirms with "Production URL has been configured."

Reviewer: ```suggestion
You can only connect production URL after the sandbox test passes — until then, the production step stays disabled. Once the test passes, in the production section of the card, click **Configure URL**. Adapty confirms with "Production URL has been configured."
```

> You can only connect production after the sandbox test passes — until then, the production step stays disabled. Once it passes, click **Configure URL** under the production setup. Adapty confirms with "Production URL has been configured."

Reviewer: не совсем stays disabled
скорее the Sandbox mode toggle is disabled while sandbox setup is incomplete
